src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
                 }
+
+                this.Reset(ref resetPoint);
+                return this.ParseSubExpression(Precedence.Expression);
             }
-            else
+            finally
             {
-                // In places where a pattern is supported, we do not support tuple types
-                // due to both syntactic and semantic ambiguities between tuple types and positional patterns.
+                this.Release(ref resetPoint);
+            }
+        }
 
-                // But it still might be a pattern such as (operand is 3) or (operand is nameof(x))
-                node = this.ParseSubExpression(Precedence.Expression);
+        private PatternSyntax ParsePattern()
+        {
+            var node = ParseExpressionOrPattern(forCase: false);
+            switch (node)
+            {
+                case PatternSyntax pattern:
+                    return pattern;
+                case ExpressionSyntax expression:
+                    return _syntaxFactory.ConstantPattern(expression);
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(node);
             }
+        }
+
+        private PropertySubpatternSyntax ParsePropertySubpattern()
+        {
+            ParseSubpatternList(out var openBraceToken, out var subPatterns, out var closeBraceToken, SyntaxKind.OpenBraceToken, SyntaxKind.CloseBraceToken);
+            return _syntaxFactory.PropertySubpattern(openBraceToken, subPatterns, closeBraceToken);
+        }
+
+        private void ParseSubpatternList(
+            out SyntaxToken openToken,
+            out Microsoft.CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList<SubpatternElementSyntax> subPatterns,
+            out SyntaxToken closeToken,
+            SyntaxKind openKind,
+            SyntaxKind closeKind)
+        {
+            Debug.Assert(openKind == SyntaxKind.OpenParenToken || openKind == SyntaxKind.OpenBraceToken);
+            Debug.Assert(closeKind == SyntaxKind.CloseParenToken || closeKind == SyntaxKind.CloseBraceToken);
+            Debug.Assert((openKind == SyntaxKind.OpenParenToken) == (closeKind == SyntaxKind.CloseParenToken));
+
+            openToken = this.EatToken(openKind);
+            var list = _pool.AllocateSeparated<SubpatternElementSyntax>();
+            try
+            {
+                if (this.CurrentToken.Kind != closeKind && this.CurrentToken.Kind != SyntaxKind.SemicolonToken)
+                {
+                    tryAgain:
+
+                    if (this.IsPossibleSubpatternElement() || this.CurrentToken.Kind == SyntaxKind.CommaToken)
+                    {
+                        // first pattern
+                        list.Add(this.ParseSubpatternElement());
 
-            return node;
+                        // additional patterns
+                        while (true)
+                        {
+                            if (this.CurrentToken.Kind == SyntaxKind.CloseParenToken ||
```

Comment:
Shouldn't this be `if (this.CurrentToken.Kind == closeKind || ... SemiColonToken`?

Reply:
No; the idea is to exit the loop even if the user wrote the wrong closing token. It keeps the parser from running away. See also `ParseArgumentList`.

---
In reply to: [151522391](https://github.com/dotnet/roslyn/pull/23208#discussion_r151522391) [](ancestors = 151522391)